source/server-selection/server-selection.rst

```diff
@@ -798,8 +798,24 @@ a pool; or (b) governed by a global or client-wide limit on number of
 waiting threads, depending on how resource limits are implemented by a
 driver.
 
-For multi-threaded clients, the server selection algorithm is
-as follows:
+operationCount
+``````````````
+
+Multi-threaded or async drivers MUST keep track of the number of operations that
+a given server is currently executing (the server's ``operationCount``). This
+value MUST be incremented once a server is selected for an operation and MUST be
+decremented once that operation has completed, regardless of its outcome. This
+value SHOULD be stored on the ``Server`` type that also owns the connection pool
+for the server, if there exists such a type in the driver's
+implementation. Incrementing or decrementing a server's ``operationCount`` MUST
+NOT wake up any threads that are waiting for a topology update as part of server
+selection. See `operationCount-based selection within the latency window
```

Comment:
When implementing this in the Rust driver, I realized that putting `operationCount` on the ServerDescription would turn server selection into a mutable operation (requiring an exclusive lock instead of a read-only lock), which seemed to not be the best option in hindsight. So in the spec, I opted to just include the soft requirement that drivers SHOULD put it on the `Server` type if they have one, and if they don't, just make sure not to wake up threads waiting in server selection when they modify the value.

Reply:
SGTM

Reply:
This sounded good to me earlier but when I implemented this in Python I realized it's much simpler to track ``operationCount`` on our Pool class rather than at the Server level:
- The pool already has checkout/checkin methods which can increment/decrement ``operationCount``.
- The pool already has a lock which can be reused for mutating ``operationCount``.
- In python, servers can be selected without actually used to run an operation (`MongoClient.primary` for example).

Putting the operationCount in the pool also transparently deals with pymongo's legacy exhaust cursor feature where a connection remains checked out until the cursor is exhausted.

Please update the guidance (and the pseudocode) to say that operationCount can live on the Pool or Server. Or perhaps we should always make it live in the pool and update the CMAP spec too?